sonarts-sq-plugin/sonar-typescript-plugin/src/main/java/org/sonar/plugin/typescript/SensorContextUtils.java

```diff
@@ -0,0 +1,227 @@
+        newIssue.addLocation(newSecondaryLocation);
+      }
+    }
+
+    if (issue.cost != null) {
+      newIssue.gap(issue.cost);
+    }
+    newIssue.save();
+  }
+
+  private static void setSecondaryLocation(NewIssueLocation newSecondaryLocation, SecondaryLocation secondaryLocation, InputFile inputFile) {
+    newSecondaryLocation.at(inputFile.newRange(
+      secondaryLocation.startLine + 1,
+      secondaryLocation.startCol,
+      secondaryLocation.endLine + 1,
+      secondaryLocation.endCol));
+    if (secondaryLocation.message != null) {
+      newSecondaryLocation.message(secondaryLocation.message);
+    }
+  }
+
+  static Iterable<InputFile> getInputFiles(SensorContext sensorContext) {
+    FileSystem fileSystem = sensorContext.fileSystem();
+    FilePredicate mainFilePredicate = sensorContext.fileSystem().predicates().and(
+      fileSystem.predicates().hasType(InputFile.Type.MAIN),
+      fileSystem.predicates().hasLanguage(TypeScriptLanguage.KEY));
+    return fileSystem.inputFiles(mainFilePredicate);
+  }
+
+  static class Issue {
+    String failure;
+    Position startPosition;
+    Position endPosition;
+    String name;
+    String ruleName;
+    SecondaryLocation[] secondaryLocations;
+    Double cost;
+
+    @Override
+    public String toString() {
+      return failure + "\n start " + startPosition + "\n end " + endPosition + "\n name " + name + "\n ruleName " + ruleName + "\n cost " + cost;
+    }
+  }
+
+  private static class Position {
+    Integer line;
+    Integer character;
+
+    @Override
+    public String toString() {
+      return "[" + line + "," + character + "]";
+    }
+  }
+
+  private static class SecondaryLocation {
+    Integer startLine;
+    Integer startCol;
+    Integer endLine;
+    Integer endCol;
+    String message;
+  }
+
+  static class AnalysisResponse {
+    String filepath;
+    Issue[] issues = {};
+    Highlight[] highlights = {};
+    CpdToken[] cpdTokens = {};
+    Symbol[] symbols = {};
+    int[] ncloc = {};
+    int[] commentLines = {};
+    Integer[] nosonarLines = {};
+    int[] executableLines = {};
+    int functions = 0;
+    int statements = 0;
+    int classes = 0;
+    int complexity = 0;
+    int cognitiveComplexity = 0;
+  }
+
+  static class Highlight {
+    Integer startLine;
+    Integer startCol;
+    Integer endLine;
+    Integer endCol;
+    String textType;
+  }
+
+  static class CpdToken {
+    Integer startLine;
+    Integer startCol;
+    Integer endLine;
+    Integer endCol;
+    String image;
+  }
+
+  static class Symbol {
+    Integer startLine;
+    Integer startCol;
+    Integer endLine;
+    Integer endCol;
+    SymbolReference[] references;
+  }
+
+  static class SymbolReference {
+    Integer startLine;
+    Integer startCol;
+    Integer endLine;
+    Integer endCol;
+  }
+
+  static class ContextualAnalysisRequest {
+    public String operation = "analyze";
+    public String file;
+    public String content;
+    public List<RuleToExecute> rules;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Make rules a static final constant or non-public and provide accessors if needed. ([squid:ClassVariableVisibilityCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=sonarts&branch=issue-351&issues=AWD5U5eJEFfjxkQcPdcl&open=AWD5U5eJEFfjxkQcPdcl)

[ISSUE_KEY]: AWD5U5eJEFfjxkQcPdcl